File: webapp/src/api/custody.ts
/**
 * The chain of custody (`docs/45`): who holds a task, where it has reached, and
 * how it fared.
 *
 * # Two clocks
 *
 * A task's *status* is what state the work is in; its *environment* is where it
 * has reached. They advance independently, which is why "resolved, on qa,
 * verified there, not yet on staging" is an ordinary sentence and no status
 * column can say it.
 *
 * # Why one read for three lists
 *
 * They are one panel and always rendered together. Three endpoints would render
 * it in three stages, and the panel's whole job is to answer "what happened to
 * this" at a glance.
 */
import { request } from './http'

export interface Transfer {
  readonly id: string
  /** `null` for the first hand-off out of triage — a transfer from nobody. */
  readonly from_team_id: string | null
  readonly to_team_id: string
  readonly moved_by: string
  readonly moved_at: string
  readonly note: string | null
}

export interface Promotion {
  readonly id: string
  readonly environment_id: string
  readonly release_id: string | null
  readonly promoted_by: string
  readonly promoted_at: string
}

export type Verdict = 'PASS' | 'FAIL'

export interface Verification {
  readonly id: string
  readonly environment_id: string
  readonly verdict: Verdict
  readonly verified_by: string
  readonly verified_at: string
  readonly note: string | null
}

export interface Custody {
  /** `null` means untriaged — not missing data, but the triage queue. */
  readonly team_id: string | null
  readonly environment_id: string | null
  readonly transfers: readonly Transfer[]
  readonly promotions: readonly Promotion[]
  readonly verifications: readonly Verification[]
}

export function readCustody(
  workspaceId: string,
  taskId: string,
  signal?: AbortSignal,
): Promise<Custody> {
  return request<Custody>(`/api/v1/tasks/${taskId}/custody`, { workspaceId, signal })
}

/**
 * Hand the task to another team.
 *
 * **This clears the assignees**, so the task lands in the receiving team's
 * queue. Any control that calls it must say so before it is pressed — a hand-off
 * that silently unassigned someone would be the worst kind of surprise.
 *
 * Refused with `409` when the task already belongs to that team, and `422`
 * (`TF-VAL-0007`) when the team is not on the task's project.
 */
export function transferTeam(
  workspaceId: string,
  taskId: string,
  teamId: string,
  note?: string,
): Promise<Transfer> {
  return request<Transfer>(`/api/v1/tasks/${taskId}/team`, {
    method: 'PUT',
    workspaceId,
    body: { team_id: teamId, ...(note === undefined || note === '' ? {} : { note }) },
  })
}

/**
 * Record that it reached an environment.
 *
 * Deliberately not idempotent: a second promotion to the same environment is a
 * redeploy, which is a real event.
 */
export function promote(
  workspaceId: string,
  taskId: string,
  environmentId: string,
): Promise<Promotion> {
  return request<Promotion>(`/api/v1/tasks/${taskId}/promotions`, {
    method: 'POST',
    workspaceId,
    body: { environment_id: environmentId },
  })
}

/**
 * Record a verdict against the environment it was tested on.
 *
 * The environment defaults to the one the task is on, which is the ordinary case
 * — QA tests what was pushed. Recording a verdict does **not** move the task;
 * what follows is a transition the caller makes next, and keeping them separate
 * is what lets "failed twice on qa" survive later status changes.
 */
export function verify(
  workspaceId: string,
  taskId: string,
  verdict: Verdict,
  note?: string,
  environmentId?: string,
): Promise<Verification> {
  return request<Verification>(`/api/v1/tasks/${taskId}/verifications`, {
    method: 'POST',
    workspaceId,
    body: {
      verdict,
      ...(note === undefined || note === '' ? {} : { note }),
      ...(environmentId === undefined ? {} : { environment_id: environmentId }),
    },
  })
}

File: crates/casual-task-persistence/src/custody.rs
//! Who holds a task, where it has reached, and whether it passed
//! (`docs/45-DEVELOPMENT-LIFECYCLE-AND-CUSTODY.md`).
//!
//! # Why these four live together
//!
//! Transfer, promotion and verification are one story told three ways: the chain
//! of custody. They are written by different handlers and read by one panel, and
//! every one of them is append-only — migration 0031 grants no `DELETE` on any
//! of these tables, because a custody chain you can edit is not a custody chain.
//!
//! # The two clocks, in code
//!
//! `task.status_id` is what state the work is in and belongs to
//! [`crate::task`]. `task.environment_id` is where it has *reached*.
//!
//! **The invariant is not "one writer" but "no silent move":** every change to
//! that column leaves a promotion row. Two functions satisfy it —
//! [`promote`], which sets the column and logs in one statement pair, and
//! [`record_promotion`], which logs a move another module already made under its
//! own optimistic-concurrency check (`crate::environment::set_on_task`, behind
//! `PUT /tasks/{id}/environment`). Advancing the second clock without a row is
//! the failure this module exists to prevent, and either function prevents it.

use time::OffsetDateTime;
use uuid::Uuid;

use crate::scoped::Scoped;

/// The tuples these rows decode into, before they become the structs above.
///
/// Named because six anonymous columns say nothing about which is which, and
/// clippy is right that the inline form is unreadable at three call sites.
type TransferTuple = (
    Uuid,
    Option<Uuid>,
    Uuid,
    Uuid,
    OffsetDateTime,
    Option<String>,
);
type PromotionTuple = (Uuid, Uuid, Option<Uuid>, Uuid, OffsetDateTime);
type VerificationTuple = (Uuid, Uuid, String, Uuid, OffsetDateTime, Option<String>);

/// One hand-off between teams.
#[derive(Debug, Clone)]
pub struct TransferRow {
    pub id: Uuid,
    /// `None` for the first assignment out of triage — a transfer from nobody.
    pub from_team_id: Option<Uuid>,
    pub to_team_id: Uuid,
    pub moved_by: Uuid,
    pub moved_at: OffsetDateTime,
    pub note: Option<String>,
}

/// One step along the second clock.
#[derive(Debug, Clone)]
pub struct PromotionRow {
    pub id: Uuid,
    pub environment_id: Uuid,
    /// Set when it moved as part of a release, `None` when a developer promoted
    /// it themselves at resolve time. Both happen, at different moments.
    pub release_id: Option<Uuid>,
    pub promoted_by: Uuid,
    pub promoted_at: OffsetDateTime,
}

/// One verdict, on one environment.
#[derive(Debug, Clone)]
pub struct VerificationRow {
    pub id: Uuid,
    pub environment_id: Uuid,
    /// `PASS` or `FAIL` — the database enum, as text.
    pub verdict: String,
    pub verified_by: Uuid,
    pub verified_at: OffsetDateTime,
    pub note: Option<String>,
}

/// Why a custody write could not be applied.
#[derive(Debug)]
pub enum CustodyError {
    /// The team is not on the task's project. `docs/45`: a task owned by people
    /// who cannot see it is not a hand-off, it is a disappearance.
    TeamNotOnProject,
    /// The environment belongs to a different project than the task.
    EnvironmentNotOnProject,
    /// The task is already owned by that team.
    AlreadyThere,
    Db(sqlx::Error),
}

impl From<sqlx::Error> for CustodyError {
    fn from(error: sqlx::Error) -> Self {
        Self::Db(error)
    }
}

impl std::fmt::Display for CustodyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::TeamNotOnProject => f.write_str("that team is not on this task's project"),
            Self::EnvironmentNotOnProject => {
                f.write_str("that environment is not on this task's project")
            }
            Self::AlreadyThere => f.write_str("the task is already owned by that team"),
            Self::Db(error) => write!(f, "{error}"),
        }
    }
}

impl std::error::Error for CustodyError {}

/// Hand a task to a team, clearing its assignees and logging the move.
///
/// # What "clearing the assignees" is for
///
/// `docs/45`: the task lands **unassigned** in the receiving team's queue, and
/// their process picks it up. Keeping the previous developer attached leaves the
/// receiving team with nothing in any queue to notice, which is the failure
/// teams complain about most. It also makes "unassigned, owned by my team" the
/// triage list a lead actually opens.
///
/// # Why the guard is a subquery and not a read
///
/// The receiving team must be on the task's project. Checked inside the same
/// statement that writes, so two concurrent requests — one removing the team
/// from the project, one transferring to it — cannot interleave into a task
/// owned by a team that has just lost its reach.
///
/// # Errors
///
/// [`CustodyError::TeamNotOnProject`], [`CustodyError::AlreadyThere`], or any
/// database error.
pub async fn transfer(
    scoped: &mut Scoped<'_>,
    task_id: Uuid,
    to_team: Uuid,
    moved_by: Uuid,
    note: Option<&str>,
) -> Result<TransferRow, CustodyError> {
    let workspace = scoped.workspace_id().as_uuid();

    let current: Option<Option<Uuid>> =
        sqlx::query_scalar("SELECT team_id FROM task WHERE id = $1 AND deleted_at IS NULL")
            .bind(task_id)
            .fetch_optional(scoped.conn())
            .await?;
    let from_team = current.ok_or(CustodyError::Db(sqlx::Error::RowNotFound))?;
    if from_team == Some(to_team) {
        return Err(CustodyError::AlreadyThere);
    }

    let moved = sqlx::query(
        "UPDATE task SET team_id = $2, updated_at = now(), updated_by = $3
          WHERE id = $1
            AND deleted_at IS NULL
            AND EXISTS (
                SELECT 1 FROM project_team pt
                 WHERE pt.project_id = task.project_id AND pt.team_id = $2)",
    )
    .bind(task_id)
    .bind(to_team)
    .bind(moved_by)
    .execute(scoped.conn())
    .await?
    .rows_affected();
    if moved == 0 {
        return Err(CustodyError::TeamNotOnProject);
    }

    // The queue is the point of the transfer, so this is not optional cleanup.
    sqlx::query("DELETE FROM task_assignee WHERE task_id = $1 AND workspace_id = $2")
        .bind(task_id)
        .bind(workspace)
        .execute(scoped.conn())
        .await?;

    let row: (
        Uuid,
        Option<Uuid>,
        Uuid,
        Uuid,
        OffsetDateTime,
        Option<String>,
    ) = sqlx::query_as(
        "INSERT INTO task_team_transfer
             (id, workspace_id, task_id, from_team_id, to_team_id, moved_by, note)
         VALUES ($1,$2,$3,$4,$5,$6,$7)
         RETURNING id, from_team_id, to_team_id, moved_by, moved_at, note",
    )
    .bind(Uuid::now_v7())
    .bind(workspace)
    .bind(task_id)
    .bind(from_team)
    .bind(to_team)
    .bind(moved_by)
    .bind(note)
    .fetch_one(scoped.conn())
    .await?;

    Ok(TransferRow {
        id: row.0,
        from_team_id: row.1,
        to_team_id: row.2,
        moved_by: row.3,
        moved_at: row.4,
        note: row.5,
    })
}

/// Move a task to an environment, recording how it got there.
///
/// Sets the column and writes the log together, so the second clock cannot
/// advance without leaving a trace of the move — see the module docs for the one
/// other path that satisfies the same invariant.
///
/// Idempotent by nature rather than by trick: promoting to the environment a
/// task is already on writes another row, because a redeploy to staging *is* a
/// second event and a log that swallowed it would understate the work.
///
/// # Errors
///
/// [`CustodyError::EnvironmentNotOnProject`] or any database error.
pub async fn promote(
    scoped: &mut Scoped<'_>,
    task_id: Uuid,
    environment_id: Uuid,
    release_id: Option<Uuid>,
    promoted_by: Uuid,
) -> Result<PromotionRow, CustodyError> {
    let workspace = scoped.workspace_id().as_uuid();

    let moved = sqlx::query(
        "UPDATE task SET environment_id = $2, updated_at = now(), updated_by = $3
          WHERE id = $1
            AND deleted_at IS NULL
            AND EXISTS (
                SELECT 1 FROM project_environment e
                 WHERE e.id = $2 AND e.project_id = task.project_id)",
    )
    .bind(task_id)
    .bind(environment_id)
    .bind(promoted_by)
    .execute(scoped.conn())
    .await?
    .rows_affected();
    if moved == 0 {
        return Err(CustodyError::EnvironmentNotOnProject);
    }

    let row: PromotionTuple = sqlx::query_as(
        "INSERT INTO task_environment_promotion
             (id, workspace_id, task_id, environment_id, release_id, promoted_by)
         VALUES ($1,$2,$3,$4,$5,$6)
         RETURNING id, environment_id, release_id, promoted_by, promoted_at",
    )
    .bind(Uuid::now_v7())
    .bind(workspace)
    .bind(task_id)
    .bind(environment_id)
    .bind(release_id)
    .bind(promoted_by)
    .fetch_one(scoped.conn())
    .await?;

    Ok(PromotionRow {
        id: row.0,
        environment_id: row.1,
        release_id: row.2,
        promoted_by: row.3,
        promoted_at: row.4,
    })
}

include!("custody_history.rs");
